github.com/cockroachdb/pebble/internal/metamorphic: TestMeta failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/6087174?buildTab=log) with [artifacts](https://buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/6087174?buildTab=artifacts#meta) on refs/heads/master @ [4c5e752da1b8](https://github.com/cockroachdb/pebble/commits/4c5e752da1b8):


```
        iter7.First() // [false] <nil> #1616
        db.Merge("jtpqgckjeh@6", "kfrg") // <nil> #1617
        iter14.SeekPrefixGE("cftvqnpwh@1") // [false] <nil> #1618
        iter10.Prev("") // [false] <nil> #1619
        batch11.Merge("hcpuftykru@4", "rlbacsli") // <nil> #1620
        iter7.SeekLT("xscyd@7", "dgwpnuqqaxr@5") // [invalid] <nil> #1621
        // INFO: [JOB 88] ingesting: sstable created 000178
        // INFO: [JOB 89] WAL created 000179
        // INFO: [JOB 90] flushing 1 memtable to L0
        // INFO: [JOB 90] flushing: sstable created 000180
        // INFO: [JOB 90] flushing: sstable created 000181
        // INFO: [JOB 90] flushing: sstable created 000182
        // INFO: [JOB 90] flushing: sstable created 000183
        // INFO: [JOB 90] flushing: sstable created 000184
        // INFO: [JOB 90] flushing: sstable created 000185
        // INFO: [JOB 90] flushing: sstable created 000186
        // INFO: [JOB 90] flushing: sstable created 000187
        // INFO: [JOB 90] flushing: sstable created 000188
        // INFO: [JOB 90] flushing: sstable created 000189
        // INFO: [JOB 90] flushing: sstable created 000190
        // INFO: [JOB 90] flushing: sstable created 000191
        // INFO: [JOB 90] flushed 1 memtable to L0 [000180 000181 000182 000183 000184 000185 000186 000187 000188 000189 000190 000191] (13 K), in 0.0s (0.0s total), output rate 10 M/s
        // INFO: [JOB 90] WAL deleted 000133
        // INFO: ingest failed to remove original file: invalid argument
        // INFO: [JOB 88] ingested L0:000178 (1.1 K)
        db.Ingest(batch11) // <nil> #1622
        iter12.SeekLT("afiovfilgy@3", "") // [false] <nil> #1623
        // INFO: [JOB 92] compacting(default) L0 [000163 000160 000157 000161 000162 000159 000156 000155 000158 000191] (11 K) + L6 [000175 000176 000177] (3.1 K)
        // INFO: [JOB 92] compacting: sstable created 000192
        db.Delete("edubjekc@7") // <nil> #1624
        db.Set("tcjkkvr@7", "mumiyeblfbpcufy") // <nil> #1625
        iter7.SeekLT("gnmg@7", "") // [false] <nil> #1626
        iter1.SeekPrefixGE("ardnvrq@6") // [false] pebble: SeekPrefixGE supplied with key outside of lower bound #1627
        iter7.SetBounds("amipb@8", "amipb@8") // <nil> #1628
        iter7.SeekLT("amipb@8", "") // [false] <nil> #1629
        // INFO: [JOB 92] compacting: sstable created 000193
        iter7.Next("") // [false] <nil> #1630
        iter7.Next("") // [false] <nil> #1631
        iter1.SetBounds("edubjekc@7", "tubikvmnsm@8") // <nil> #1632
        // INFO: [JOB 92] compacting: sstable created 000194
        iter1.SeekGE("edubjekc@7", "") // [false] <nil> #1633
        // INFO: [JOB 92] compacting: sstable created 000195
        iter1.Prev("") // [false] <nil> #1634
        iter1.Prev("") // [false] <nil> #1635
        iter6.SeekLT("tuyvf@1", "") // [false] <nil> #1636
        // INFO: [JOB 92] compacting: sstable created 000196
        snap9.Get("cftvqnpwh") // [""] pebble: not found #1637
        iter15 = iter7.Clone(true) // <nil> #1638
        // INFO: [JOB 92] compacting: sstable created 000197
    --- FAIL: TestMeta/standard-015 (0.13s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -mod=vendor -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run TestMeta$ ./internal/metamorphic -seed 1660455701748200263 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

```console
=== RUN   TestMeta/standard-015
    meta_test.go:327:
        ===== SEED =====
        1660455701748200263
        ===== ERR =====
        exit status 2
        ===== OUT =====
        --- FAIL: TestMeta (0.07s)
        panic: pebble: invariant violation: no next span while switching directions [recovered]
                panic: pebble: invariant violation: no next span while switching directions

        goroutine 21 [running]:
        testing.tRunner.func1.2({0x45a31c0, 0x470b320})
                /nix/store/v6hllh9fg47llmbqa8c88ihcbpxr4cjv-go-1.19/share/go/src/testing/testing.go:1396 +0x24e
        testing.tRunner.func1()
                /nix/store/v6hllh9fg47llmbqa8c88ihcbpxr4cjv-go-1.19/share/go/src/testing/testing.go:1399 +0x39f
        panic({0x45a31c0, 0x470b320})
                /nix/store/v6hllh9fg47llmbqa8c88ihcbpxr4cjv-go-1.19/share/go/src/runtime/panic.go:884 +0x212
        github.com/cockroachdb/pebble/internal/keyspan.(*DefragmentingIter).Next(0xc0002ecc30)
                /Users/nickt/Development/pebble/internal/keyspan/defragment.go:268 +0x16b
        github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).keyspanSeekGE(0xc0002ec628, {0xc00041a1c0, 0x8, 0x8}, {0xc0001ebc60, 0x9, 0x9})
                /Users/nickt/Development/pebble/internal/keyspan/interleaving_iter.go:763 +0xc4
        github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).InitSeekGE(0xc0002ec628, {0xc0001ebc60, 0x9, 0x9}, {0xc00041a1c0, 0x8, 0x8}, 0xc000217408?, {0x6000021180c2, 0x1e, ...})
                /Users/nickt/Development/pebble/internal/keyspan/interleaving_iter.go:218 +0x14f
        github.com/cockroachdb/pebble.(*lazyCombinedIter).initCombinedIteration(0xc000445730, 0x1, 0xc000217408, {0x6000021180c2, 0x1e, 0x1e}, {0x0, 0x0, 0x0})
                /Users/nickt/Development/pebble/range_keys.go:532 +0x4f0
        github.com/cockroachdb/pebble.(*lazyCombinedIter).Next(0xc000445730)
                /Users/nickt/Development/pebble/range_keys.go:600 +0x65
        github.com/cockroachdb/pebble.(*Iterator).nextUserKey(0xc000445400)
                /Users/nickt/Development/pebble/iterator.go:574 +0x18f
        github.com/cockroachdb/pebble.(*Iterator).NextWithLimit(0xc000445400, {0x0, 0x0, 0x0})
                /Users/nickt/Development/pebble/iterator.go:1370 +0x4f1
        github.com/cockroachdb/pebble.(*Iterator).Next(...)
                /Users/nickt/Development/pebble/iterator.go:1339
        github.com/cockroachdb/pebble/internal/metamorphic.(*retryableIter).Next.func1()
                /Users/nickt/Development/pebble/internal/metamorphic/retryable.go:134 +0x35
        github.com/cockroachdb/pebble/internal/metamorphic.(*retryableIter).withRetry(0xc0002953e0, 0xc0005a76a8)
                /Users/nickt/Development/pebble/internal/metamorphic/retryable.go:61 +0x31
        github.com/cockroachdb/pebble/internal/metamorphic.(*retryableIter).Next(0x0?)
                /Users/nickt/Development/pebble/internal/metamorphic/retryable.go:133 +0x4d
        github.com/cockroachdb/pebble/internal/metamorphic.(*iterNextOp).run(0xc0001fb200, 0xc0000ee790?, 0xc000330570?)
                /Users/nickt/Development/pebble/internal/metamorphic/ops.go:918 +0xa7
        github.com/cockroachdb/pebble/internal/metamorphic.(*test).step(0xc0000ee790, 0xc0000a3ba0?)
                /Users/nickt/Development/pebble/internal/metamorphic/test.go:206 +0x48
        github.com/cockroachdb/pebble/internal/metamorphic.testMetaRun(0xc000080cf0?, {0x7ff7bfeff51f, 0x24}, 0x4b895b8?, {0xc0000e48a0, 0x2c})
                /Users/nickt/Development/pebble/internal/metamorphic/meta_test.go:194 +0xa7a
        github.com/cockroachdb/pebble/internal/metamorphic.TestMeta(0xc0000a3ba0)
                /Users/nickt/Development/pebble/internal/metamorphic/meta_test.go:241 +0x8b6
        testing.tRunner(0xc0000a3ba0, 0x4682b28)
                /nix/store/v6hllh9fg47llmbqa8c88ihcbpxr4cjv-go-1.19/share/go/src/testing/testing.go:1446 +0x10b
        created by testing.(*T).Run
                /nix/store/v6hllh9fg47llmbqa8c88ihcbpxr4cjv-go-1.19/share/go/src/testing/testing.go:1493 +0x35f
```

That test variant is

```
[Level "0"]
  target_file_size=1
```

--- a/internal/keyspan/bounded.go
+++ b/internal/keyspan/bounded.go
@@ -32,6 +32,22 @@ const (
 )
 
 // BoundedIter implements FragmentIterator and enforces bounds.
+//
+// Like the point InternalIterator interface, the bounded iterator's forward
+// positioning routines (SeekGE, First, and Next) only check the upper bound.
+// The reverse positioning routines (SeekLT, Last, and Prev) only check the
+// lower bound. It is up to the caller to ensure that the forward positioning
+// routines respect the lower bound and the reverse positioning routines respect
+// the upper bound (i.e. calling SeekGE instead of First if there is a lower
+// bound, and SeekLT instead of Last if there is an upper bound).
+//
+// When the hasPrefix parameter indicates that the iterator is in prefix
+// iteration mode, BoundedIter elides any spans that do not overlap with the
+// prefix's keyspace. In prefix iteration mode, reverse iteration is disallowed,
+// except for an initial SeekLT with a seek key greater than or equal to the
+// prefix. In prefix iteration mode, the first seek must position the iterator
+// at or immediately before the first fragment covering a key greater than or
+// equal to the prefix.
 type BoundedIter struct {
 	iter      FragmentIterator
 	iterSpan  *Span

--- a/iterator.go
+++ b/iterator.go
@@ -1351,10 +1351,20 @@ func (i *Iterator) Next() bool {
 // keyspace up to limit.
 func (i *Iterator) NextWithLimit(limit []byte) IterValidityState {
 	i.stats.ForwardStepCount[InterfaceCall]++
-	if limit != nil && i.hasPrefix {
-		i.err = errors.New("cannot use limit with prefix iteration")
-		i.iterValidityState = IterExhausted
-		return i.iterValidityState
+	if i.hasPrefix {
+		if limit != nil {
+			i.err = errors.New("cannot use limit with prefix iteration")
+			i.iterValidityState = IterExhausted
+			return i.iterValidityState
+		} else if i.iterValidityState == IterExhausted {
+			// No-op, already exhasuted. We avoid executing the Next because it
+			// can break invariants: Specifically, a file that fails the bloom
+			// filter test may result in its level being removed from the
+			// merging iterator. The level's removal can cause a lazy combined
+			// iterator to miss range keys and trigger a switch to combined
+			// iteration at a larger key, breaking keyspan invariants.
+			return i.iterValidityState
+		}
 	}
 	if i.err != nil {
 		return i.iterValidityState

--- a/testdata/iterator
+++ b/testdata/iterator
@@ -880,8 +880,8 @@ next
 a:bc
 .
 .
-stats: (interface (dir, seek, step): (fwd, 1, 2), (rev, 0, 0)), (internal (dir, seek, step): (fwd, 1, 4), (rev, 0, 0)),
-(internal-stats: (block-bytes: (total 0 B, cached 0 B)), (points: (count 12, key-bytes 16, value-bytes 12, tombstoned: 0))
+stats: (interface (dir, seek, step): (fwd, 1, 2), (rev, 0, 0)), (internal (dir, seek, step): (fwd, 1, 2), (rev, 0, 0)),
+(internal-stats: (block-bytes: (total 0 B, cached 0 B)), (points: (count 8, key-bytes 10, value-bytes 8, tombstoned: 0))
 
 iter seq=2
 seek-prefix-ge a
@@ -891,8 +891,8 @@ next
 a:b
 .
 .
-stats: (interface (dir, seek, step): (fwd, 1, 2), (rev, 0, 0)), (internal (dir, seek, step): (fwd, 1, 2), (rev, 0, 0)),
-(internal-stats: (block-bytes: (total 0 B, cached 0 B)), (points: (count 14, key-bytes 18, value-bytes 14, tombstoned: 0))
+stats: (interface (dir, seek, step): (fwd, 1, 2), (rev, 0, 0)), (internal (dir, seek, step): (fwd, 1, 1), (rev, 0, 0)),
+(internal-stats: (block-bytes: (total 0 B, cached 0 B)), (points: (count 10, key-bytes 14, value-bytes 10, tombstoned: 0))
 
 iter seq=4
 seek-prefix-ge a

